### ISSUE TITLE
jax.tree_util.Partial is not hash-stable in jax>=0.2.22

### ISSUE THREAD

**Issue by PhilipVinc** [google/jax#9429]
Following #8101 , `jax.tree_util.Partial.func` is no more hash-stable:

jax==0.2.21
```python
>>> import jax
>>> from jax.tree_util import Partial
>>> from functools import partial
>>> fun = partial(print, 1)
>>> hash(fun)
278539187
>>> hash(Partial(fun, 2).func)
278539187
>>> hash(Partial(fun, 2).func)
278539187
```

latest release
```python
>>> from jax.tree_util import Partial
>>> from functools import partial
>>> fun = partial(print, 1)
>>> hash(fun)
8749850128694
>>> hash(Partial(fun, 2).func)
8750211514691
>>> hash(Partial(fun, 2).func)
8750211514664

# instead, if the function wrapped is just a standard function, everything works fine
>>> hash(Partial(print, 1).func)
56249262
>>> hash(Partial(print, 1).func)
56249262
```

This is annoying because in some code in NetKet, we were wrapping user-passed functions into a `tree_util.Partial` in order to simplify some code, but now this triggers recompilation if the user's supplied function is wrapped in a partial or not.

cc @hawkinsp who committed the changes

**Comment by hawkinsp** [comment]
#9430 fixes the problem identified in this bug. Note that `.func` will still not return the original object: instead it returns a wrapper object that has the same hash and compares as equal to the original function. If we wanted to avoid that, we'd need to reimplement `Partial` without subclassing `functools.partial`.

**Comment by PhilipVinc** [comment]
That's great, thanks and solves our issue!

I mainly care about the hash stability so that I can use that without triggering recompilation. The fact that the object is not identical does not seem a problem to me.

**Comment by hawkinsp** [comment]
We released `jax` 0.3.0 which includes this fix.
